h1b/src/nvcounter/capsule.rs

```diff
@@ -0,0 +1,293 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use ::kernel::ReturnCode;
+use super::internal::*;
+use super::traits::{Client,NvCounter};
+use crate::hil;
+
+/// NvCounter implementation using flash memory.
+
+// Uses two pages of flash, flipping two bits to zero in each page at a time.
+// The low page is struck or erased for every increment operation. When the low
+// page wraps around, the high page is struck once, the low page is erased, then
+// the high page is struck again. This allows the counter to resume if the
+// increment operation is interrupted, providing atomicity.
+//
+// initialize() sequence:
+//   A1. Erase low
+//   A2. Erase high
+//
+// Normal increment() sequence:
+//   B1. Write low
+//
+// Rollover increment() sequence:
+//   C1. Write high (even -> odd)
+//   C2. Erase low [in background]
+//   C3. Write high (odd -> even) [in background]
+//
+// Possible values of task, high page count, and low page count when each step
+// completes, before the task value is updated:
+//   Step  Task        High  Low
+//   A1    Initialize  *     0
+//   A2    Initialize  0     0
+//   B1    Increment   Even  >0
+//   C1    Increment   Odd   Max
+//   C2    *           Odd   0
+//   C3    *           Even  0
+
+pub struct FlashCounter<'c, F: hil::flash::Flash<'c> + 'c> {
+    client: ::core::cell::Cell<Option<&'c dyn Client>>,
+    flash: &'c F,
+    write_buffer: core::cell::Cell<Option<&'c mut [u32; 1]>>,
+
+    // What operation the client is currently waiting on. Note that when
+    // executing steps C2 and C3, this reflects the task the client wants the
+    // counter to do *next*.
+    task: ::core::cell::Cell<Option<Task>>,
+}
+
+impl<'c, F: hil::flash::Flash<'c> + 'c> FlashCounter<'c, F> {
+    pub fn new(buffer: &'c mut [u32; 1], flash: &'c F) -> Self {
+        FlashCounter {
+            client: ::core::cell::Cell::new(None),
+            flash,
+            write_buffer: core::cell::Cell::new(Some(buffer)),
+            task: ::core::cell::Cell::new(None),
+        }
+    }
+}
+
+impl <'c, F: hil::flash::Flash<'c> + 'c> NvCounter<'c> for FlashCounter<'c, F> {
+    fn initialize(&self) -> ReturnCode {
+        // For now, we only support doing a single operation at a time.
+        if self.task.get().is_some() { return ReturnCode::EBUSY; }
+        // Try to start the erase (step A1). If a flash operation is ongoing, we
+        // will get back EBUSY. In that case, return success, as the erase will
```

Comment:
Under what circumstances should EBUSY be returned from `erase` when there is no task? If there's broken virtualization? Or is the driver assuming no virtualization? If so, will it get callbacks (e.g., is it the client)?

Reply:
I added an explanation. EBUSY will be returned when a rollover increment is being finished in the background.